internal/queue/queue.go

```diff
@@ -247,3 +280,104 @@ func NewQueue(discovered discovery.DiscoveredConfigs) (*Queue, error) {
 
 	return q, errors.New("cycle detected during queue construction")
 }
+
+// GetReadyWithDependencies returns all entries that are ready to run and have all dependencies completed(or no dependencies).
+func (q *Queue) GetReadyWithDependencies() []*Entry {
+	out := make([]*Entry, 0, len(q.Entries))
+
+	for _, e := range q.Entries {
+		if e.Status == StatusReady {
+			allDepsReady := true
+
+			for _, dep := range e.Config.Dependencies {
+				depEntry, ok := q.Index[dep.Path]
+				if !ok || depEntry.Status != StatusSucceeded {
+					allDepsReady = false
+					break
+				}
+			}
+
+			if allDepsReady {
+				out = append(out, e)
+			}
+		}
+	}
+
+	return out
+}
+
+// SetStatus sets the status of the given entry, handles fail-fast logic if enabled, and updates the statuses of dependents accordingly.
+// This method should be used for all status transitions. If fail-fast is enabled and the entry fails, all not-yet-started tasks are marked as failed.
+// It also unblocks children or marks them as failed/ancestor-failed as appropriate.
+func (q *Queue) SetStatus(e *Entry, status Status) {
+	e.Status = status
+
+	// If this entry failed, recursively fail all dependents using the Dependents field
+	if status == StatusFailed {
+		// If fail-fast is enabled, mark all not-yet-started tasks (Pending, Blocked, Ready) as Failed to prevent further execution.
+		if q.FailFast {
+			for _, n := range q.Entries {
+				if isTerminalOrRunning(n.Status) {
+					continue
+				}
+
+				n.Status = StatusFailed
```

Comment:
Should this be a call to `q.SetStatus(n, StatusFailed)`? That would make it recursive, I believe. See my comment above about whether this should change to "Propagate Failure"

Also, see my comment above about this potentially being better off as `StatusEarlyExit` to stay consistent with reporting.